dali/pipeline/util/bounding_box.h

```diff
@@ -20,170 +20,382 @@
 #include <vector>
 #include <limits>
 #include <string>
-
 #include "dali/core/error_handling.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/tensor_layout.h"
+#include "dali/core/format.h"
+#include "dali/core/math_util.h"
 
 namespace dali {
 
+using RelCoords = SmallVector<float, 3>;
+using RelBounds = SmallVector<float, 6>;  // e.g. start0, start1, ..., end0, end1, ...
+
 class BoundingBox {
  public:
-  static const size_t kSize = 4;
-  static constexpr array<float, kSize> NoBounds() {
-    return {
-      std::numeric_limits<float>::lowest(),
-      std::numeric_limits<float>::lowest(),
-      std::numeric_limits<float>::max(),
-      std::numeric_limits<float>::max()};
+  static RelBounds Uniform(int ndim = 2, float min = 0.0f, float max = 1.0f) {
```

Comment:
I stand by the opinion that this is not the best design choice to have dimensionality of the boxes defined at run-time. I think this class should be a `template <int ndim>` and all of the dimensionality checks would be simply unnecessary.
Also, it would go back to being standard layout type, enabling it to be usable on device-side. Moreover, we already have most of the logic for handling fixed-dim boxes in core/geom/box.h.